---
title: "Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


### Import countData and colData
```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = F)
metadata <- read.csv("airway_metadata.csv", stringsAsFactors = F)
```

How many genes are in this database?
```{r}
nrow(counts)
```
How many experiments?
```{r}
ncol(counts)
```

Let's check for metadata to counts correspondence
```{r}
all( metadata$id == colnames(counts)[-1] )
```

Extract controls and calculate mean
```{r}
# Find the column names (ids) of the 'control' experiments
control <- metadata[metadata$dex=="control",]

# Use the ids to extract the control counts columns
control.mean <- rowSums(counts[, control$id]) / nrow(control)
names(control.mean) <- counts$ensgene
```

Extract treated and calculate mean
```{r}
treated <- metadata[metadata$dex=="treated",]
treated.mean <- rowSums(counts[, treated$id]) / nrow(treated)
names(treated.mean) <- counts$ensgene
```

Combine mean counts and check sums
```{r}
meancounts <- data.frame(control.mean, treated.mean)
colSums(meancounts)
```

Plot mean counts
```{r}
plot(control.mean, treated.mean, log="xy", xlab="Control", ylab="Treated")
```

Calculate log2 fold changes of the mean counts
```{r}
meancounts$log2fc <- log2( meancounts$treated.mean / meancounts$control.mean )
```

Filter out weird values
```{r}
zero.vals <- which(meancounts[,1:2] == 0, arr.ind = T)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
```

Check for + - 2 threshold
```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)

sum(up.ind)
sum(down.ind)
```

Import annotations
```{r}
anno <- read.csv("annotables_grch38.csv")
```

Merge annotations with my counts
```{r}
anno.mycounts <- merge(mycounts, anno, by.x = "row.names", by.y = "ensgene")
```


```{r}
# biocLite("AnnotationDbi")
# biocLite("org.Hs.eg.db")
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db, 
                          keys = row.names(mycounts),
                          column = "SYMBOL",
                          keytype = "ENSEMBL",
                          multiVals = "first")
```

```{r}
mycounts$entrez <- mapIds(org.Hs.eg.db, 
                          keys = row.names(mycounts),
                          column = "ENTREZID",
                          keytype = "ENSEMBL",
                          multiVals = "first")
```

```{r}
mycounts$uniprot <- mapIds(org.Hs.eg.db, 
                          keys = row.names(mycounts),
                          column = "UNIPROT",
                          keytype = "ENSEMBL",
                          multiVals = "first")
```

### DESeq 2 analysis
```{r}
# source("http://bioconductor.org/biocLite.R")
# biocLite()

# biocLite("DESeq2")
library("DESeq2")
```

Build our DESeq dataset required for DESeq analysis
```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design = ~dex,
                              tidy = T)
dds
```

Run the DESeq analysis
```{r}
dds <- DESeq(dds)
```

Get our results
```{r}
res <- results(dds)
res
```

```{r}
summary(res)
```


Make volcano plot
```{r}
mycols <- rep("gray", nrow(res))
mycols[abs(res$log2FoldChange) > 2] <-"red"
subset.inds <- res$padj < 0.01 & abs(res$log2FoldChange) > 2
mycols[subset.inds] <- "blue"
plot(res$log2FoldChange, -log(res$padj), col=mycols)
```

```{r}

```

